### ISSUE TITLE
Nested Resampling is not like described in the paper

### ISSUE THREAD

**Issue by FlorianPargent** [sebastiansauer/reanalysis-mindful-machine-learning#1]
Thank you for sharing the reanalysis with me!

Currently, the resampling scheme in the reanalysis does not reflect what is described in the paper:

> We employed a more stringent variant known as the repeated k-fold cross-validation procedure (Kuhn, 2008). Here, the data are split into k distinct blocks of similar size. In our case, we split the data into k = 10 blocks and repeated this procedure for n = 10 repetitions. Then, in n runs, each of the k blocks was left out, and the model was fit to the remaining 9/10 of the sample. The model results, built on the “left out” 1/10, were then averaged.

To me, this implies 10 times repeated 10 fold cross-validation for performance evaluation.
The current reanalysis uses a simple holdout strategy (train model on a single training set and evaluate on a single test set). The repeated cross-validation mentioned in the code is instead used for hyperparameter tuning in the training set (in my opinion simple cross-validation should be enough here). The original manuscript does not mention how hyperparameter tuning was performed (as a reader I thought it was not done at all), although now I suppose tuning was done implicitly by caret (as is shown in your current code). When combining hyperparameter tuning with performance evaluation, nested resampling is needed (and correctly employed in the current example for a simple holdout set). According to my knowledge, this cannot be fully automated in caret when using more complicated resampling strategies for the outer performance evaluation loop (in contrast to other meta machine learning packages). However, I drafted some code (should be checked by you as I'm still not an expert on caret) how this could be done with the existing caret functionality for the example of the random forest algorithm (currently using the caret default hyper parameter grid options).

My example uses 10 fold CV for hyperparameter tuning (in my oppinion a reasonable choice) in the inner resampling loop and 10 times repeated 10 fold CV for performance evaluation in the outer resampling loop:

```
library(caret)
set.seed(998)

folds <- createMultiFolds(dat2$practice, k = 10, times = 10) ## 10 rep 10-fold CV for performance evaluation

inner_res <- trainControl(method = "cv", number = 10) ## 10-fold CV for hyperparameter tuning

mod_list <- lapply(folds, function(x) { # train models for each of the 100 training sets
  train(practice ~ ., data = dat2[x,], 
                 method = "rf", 
                 trControl = inner_res,
                 verbose = FALSE)
})

pred_list <- mapply(function(x, y) { # compute predictions on each of the 100 test sets
  predict(x, dat2[-y,])}, 
  x = mod_list, y = folds)

perf_sens_spec_list <- mapply(function(x, y) { # calculate performance measures I
  confusionMatrix(x, dat2[-y, "practice", drop = TRUE])$byClass[c(1,2)]}, 
  x = pred_list, y = folds)

perf_accuracy_kappa_list <- mapply(function(x, y) { # calculate performance measures II
  confusionMatrix(x, dat2[-y, "practice", drop = TRUE])$overall[c(1,2)]}, 
  x = pred_list, y = folds)

# average performance across the 100 test sets
mean(perf_sens_spec_list["Sensitivity", ])
mean(perf_sens_spec_list["Specificity", ])
mean(perf_accuracy_kappa_list["Accuracy", ])
mean(perf_accuracy_kappa_list["Kappa", ])

``` 

Just ask if you have any questions!

**Comment by sebastiansauer** [comment]
Thanks for this note and the code snippet!

>    Currently, the resampling scheme in the reanalysis does not reflect what is described in the paper:


I agree that our description is not crystal clear, and I admit that I only used to know the "caret way" of resampling. However, I did not state that we did nested resampling and I cited a source where the details of the method applied is spelled out (Max Kuhn is the author of the caret package).

That said, I agree that the idea of nesting resampling is great. For some reason, Max Kuhn seems not to share this opinion, see [this interesting SO post](https://stats.stackexchange.com/questions/378426/parameter-tuning-with-vs-without-nested-cross-validation). I can't really follow Max' concerns, but tend to think that the "nested" approach is a good idea.

I've detailled the caret way of resampling [here](https://github.com/sebastiansauer/reanalysis-mindful-machine-learning/blob/master/reanalysis-code/comment-nested-sampling.Rmd), if you are interested. We used the caret functions, so the caret way is what is implemented in the paper.

What I suggest is the following: I will implement your approach (and I'll try write it quite general, so that it may be of use elsewhere). Roughly speaking, your approach consists of putting an "outer loop" around what I implemented, if I'm not mistaken. That's what I'm planning to build.

But I would not recommend it putting it to the original paper as we would be "rewriting history".

**Comment by FlorianPargent** [comment]
Thank you for your detailed explanation!
The "caret way" of resampling is completely fine in the methodological sense that it produces a conservative performance evaluation of the final model due to the independent test set, although for small sample sizes, both bias and variance of the performance estimator based on a single training/test split can be expected to be large.
I totally understand that you do not want to implement nested resampling for the original paper, as this is a lot of work when using caret.

I still think that your original method description ...

> One of the dangers of data modeling is overfitting. Overfitting refers to a situation in which a statistical model is excessively complex and thereby describes random noise as a systematic pattern. **A possible solution to this problem is to split the data into a training set and a test set (Hastie et al., 2009). The model is then built on the training set only; for model results, the test set is used. We employed a more stringent variant known as the repeated k-fold cross-validation procedure (Kuhn, 2008).** Here, the data are split into k distinct blocks of similar size. In our case, we split the data into k = 10 blocks and repeated this procedure for n = 10 repetitions. Then, in n runs, each of the k blocks was left out, and the model was fit to the remaining 9/10 of the sample. The model results, built on the “left out” 1/10, were then averaged.

where you stress that repeated k-fold cross-validation is a more stingent variant of the training test split is somewhat misleading. With the employed resampling strategy, the only reason for introducing the repeated cross-validation in the training set is hyperparameter tuning (without some resampling strategy applied on the training set, hyper parameter tuning would not be possible, as there would be no reasonable way to decide which tuning configuration is expected to yield the best performance). The quality of performance evaluation is not improved as the performance of the final model is still evaluated on a single testset, similar to the simple training test method (without any hyperparameter tuning) described in Hastie et al. (2009) and others.

However, if you do not want to change the description of your resampling scheme I can live with that.

Thanks again for the professional exchange!

**Comment by sebastiansauer** [comment]
Dear Florian,

Thank your for your comment. I  agree that we could (and maybe should) have been more explicit in the description of our method as it comes to the pecularities of the cross-validation method. 

I also agree that our method (ie., Kuhn's/caret's method) amounts to using cross-validation *only for hyper parameter selection*, but *not* for test set validation. The caret method leaves us with a simple train-test set split which I know see as not ideal. When I/we wrote the paper back in 2015 I only knew of Kuhn's method and just followed it. Interestingly, in the post I cited above, Kuhn seems to still favor his method. It is by your comment that I now understand (or at least believe so) that the "full" or "nested" approach is preferable (in many but not all situations). 

Following your example, I have implemented the nested approach in the reanalyis code (it's on Github, but as to now only at 95% ready). I will note you as soon as it is finished, and would be interested in knowing whether that's similar to your code. Basically, I've added an "outer loop" around my code.

As for the present paper and the two points you raised, I'd suggest the following: I'll contact the editor (cc-ing you) stating

1.  that I (the authors) of the paper mistakenly used the train set for prediction where the test set would reveal more reliable results. This is a mistake in need for correction.
2. that we could have been more explicit in describing our method, and that some readers may be mislead. I'll recommend to the editor that I rephrase some sentences in the method section accordingly.

I hope that your concerns are adequately addressed in this way. Note that all materials (including the original code with the time stamp of 2015) ore openly available in this repo. Please let me know if you feel that any of your concerns still await consideration. Again, let me thank you for pointing out a flaw in our analysis I was'nt aware of and only learnt to see it by your remark.

**Comment by sebastiansauer** [comment]
Here's the link to the updated re-analyses, see "nested resampling" section:https://github.com/sebastiansauer/reanalysis-mindful-machine-learning/blob/master/Reanalysis-Mindful-Machine_Learning_01.Rmd

https://github.com/sebastiansauer/reanalysis-mindful-machine-learning/blob/master/Reanalysis-Mindful-Machine_Learning_01.html
